aesara/tensor/blockwise.py

```diff
@@ -0,0 +1,250 @@
+
+    """
+    return [
+        broadcast_shape + tuple(dim_sizes[dim] for dim in core_dims)
+        for core_dims in list_of_core_dims
+    ]
+
+
+def gufunc_sign_to_str(sign):
+    in_sign = [f"({','.join(_sign)})" for _sign in sign[0]]
+    out_sign = [f"({','.join(_sign)})" for _sign in sign[1]]
+    return f"{','.join(in_sign)}->{','.join(out_sign)}"
+
+
+class Blockwise(Op):
+    __props__ = ("op", "signature")
+
+    def __init__(self, op, signature=None):
+        self.op = op
+        self.signature = signature or self.op.gufunc_sig
+
+    def make_node(self, *inputs):
+
+        num_expected_inps = len(self.signature[0])
+        if len(inputs) != num_expected_inps:
+            raise ValueError(
+                f"Expected {int(num_expected_inps)} inputs, got {len(inputs)}"
+            )
+
+        # TODO: Correct this
+        out_dtype = inputs[0].dtype
+
+        bcast_shape, dim_sizes = _parse_input_dimensions(inputs, self.signature[0])
+        output_shapes = _calculate_shapes(bcast_shape, dim_sizes, self.signature[1])
+
+        def safe_const_val(x):
+            try:
+                return get_scalar_constant_value(x)
+            except NotScalarConstantError:
+                return None
+
+        outputs = [
+            TensorType(out_dtype, shape=tuple(safe_const_val(s) for s in shp))()
+            for shp in output_shapes
+        ]
+        return Apply(self, list(inputs), outputs)
+
+    def infer_shape(self, fgraph, node, shapes):
+        bcast_shape, dim_sizes = _parse_input_dimensions(node.inputs, self.signature[0])
+        output_shapes = _calculate_shapes(bcast_shape, dim_sizes, self.signature[1])
+        return output_shapes
+
+    def L_op(
+        self,
+        inputs: Sequence[Variable],
+        outputs: Sequence[Variable],
+        ograds: Sequence[Variable],
+    ):
+
+        with aesara.config.change_flags(compute_test_value="off"):
+            core_inputs = []
+            for _inp, _inp_sig in zip(inputs, self.signature[0]):
+                curr_dtype = _inp.type.dtype
+                curr_static_shape = _inp.type.shape[: len(_inp_sig)]
+                core_inputs.append(TensorType(curr_dtype, curr_static_shape)())
+
+            core_out_grads = []
+            for _out_grad, _out_sig in zip(ograds, self.signature[1]):
+                curr_dtype = _out_grad.type.dtype
+                curr_static_shape = _out_grad.type.shape[: len(_out_sig)]
+                core_out_grads.append(TensorType(curr_dtype, curr_static_shape)())
+
+            core_outputs: Sequence[Variable] = self.op.make_node(*core_inputs).outputs
+            core_inp_grads = self.op.L_op(core_inputs, core_outputs, core_out_grads)
+
+            for igrad in core_inp_grads:
+                assert igrad is not None, self.op
+
+        def transform(var: "TensorVariable", client_node: Optional[Apply]) -> Variable:
+            """Walk a graph and expand single gradient \"block\"s into their block-wise equivalents."""
```

Comment:
Hi @brandonwillard 
Can you explain what transform function is and how it is used in computing L_Op?

Reply:
Just like its `Elemwise` counterpart, `transform` is supposed to use a "template" gradient graph for each input to construct a real gradient graph in which all the relevant `Op`s are `Elemwise`/`Blockwise` `Op`s applied to the original inputs.

Let's take a look at what's happening in `Blockwise.L_op` in first `test_Blockwise_grad` test.

First, the graph for which we want the L-op/gradient:
```python
aesara.dprint(outputs)
# Blockwise{op=<tests.tensor.test_blockwise.DotBW object at 0x7f5e8236fd90>, signature=((('m', 'n'), ('n', 'p')), (('m', 'p'),))} [id A] <TensorType(float64, (None, None, None))>
#  |input 0 [id B] <TensorType(float64, (None, None, None))>
#  |input 1 [id C] <TensorType(float64, (None, None, None))>
```
It's a `Blockwise` dot product node with two 3D inputs named `input 0` and `input 1`.

A "template" graph of the gradient is produced for each input and stored in `core_inp_grads`.  Each element of `core_inp_grads` corresponds to the generic form of a single-block's gradient wrt. each input.
```python
aesara.dprint(core_inp_grads, print_type=True)
# dot [id A]
#  |<TensorType(float64, (None, None))> [id B]
#  |InplaceDimShuffle{1,0} [id C]
#    |<TensorType(float64, (None, None))> [id D]
# dot [id E]
#  |InplaceDimShuffle{1,0} [id F]
#  | |<TensorType(float64, (None, None))> [id G]
#  |<TensorType(float64, (None, None))> [id B]
```
We can see that the gradient of a `dot` in a single block is just another `dot`.  The original inputs aren't present; instead some stand-in variables are used and they're 2D (i.e. `TensorType`s with `(None, None)` static shapes).  
In other words, we've used the core dimensions specified by the `Blockwise` and its `Op` to remove the broadcasted dimensions (i.e. that determine each block) and produce the generic form of a single "block"'s L-op from an existing `Op.[grad|L_op]` implementation.

Now, we can't simply replace those stand-in inputs with `input 0` and/or `input 1`, because the `dot` in the gradient graphs doesn't work block-wise and, as a result, cannot take the original inputs as inputs.  Also, the `InplaceDimShuffle` applied to one of the inputs in each graph wouldn't work with an input containing an extra dimension.

The idea is that we need to convert the template's `dot` to a `Blockwise(dot)` and do something about the `InplaceDimShuffle`s.

This is a very specific example, but the general idea is that we need to read these template gradient graphs for a single block and convert all the relevant nodes into block-wise operations that are applied to the original inputs.